src/client-side-metrics/operation-metrics-collector.ts

```diff
@@ -139,76 +209,92 @@ export class OperationMetricsCollector {
+                    attemptLatency: totalTime,
+                    serverLatency: this.serverTime ?? undefined,
+                    connectivityErrorCount: this.connectivityErrorCount,
+                    streaming: this.streamingOperation,
+                    status: attemptStatus.toString(),
+                    client_name: `nodejs-bigtable/${version}`,
+                    metricsCollectorData: this.getMetricsCollectorData(),
+                    projectId,
+                  });
+                }
+              }
+            );
+          } else {
+            console.warn('ProjectId and start time should always be provided');
           }
-        });
-      }
-    } else {
-      console.warn('Invalid state transition attempted');
-    }
+        }
+      );
+    });
   }
 
   /**
    * Called when a new attempt starts. Records the start time of the attempt.
    */
   onAttemptStart() {
-    if (
-      this.state ===
-      MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS
-    ) {
-      this.state =
-        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET;
-      this.attemptStartTime = new Date();
-      this.serverTime = null;
-      this.serverTimeRead = false;
-      this.connectivityErrorCount = 0;
-    } else {
-      console.warn('Invalid state transition attempted');
-    }
+    withMetricsDebug(() => {
+      checkState(
+        this.state,
+        [MetricsCollectorState.OPERATION_STARTED_ATTEMPT_NOT_IN_PROGRESS],
+        () => {
+          this.state =
+            MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET;
+          this.attemptStartTime = new Date();
+          this.serverTime = null;
+          this.serverTimeRead = false;
+          this.connectivityErrorCount = 0;
+        }
+      );
+    });
   }
 
   /**
    * Called when the first response is received. Records first response latencies.
    */
   onResponse(projectId: string) {
-    if (!this.firstResponseLatency) {
-      // Check firstResponseLatency first to improve latency for calls with many rows
-      if (
-        this.state ===
-        MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET
-      ) {
-        this.state =
-          MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED;
-        const endTime = new Date();
-        if (projectId && this.operationStartTime) {
-          this.firstResponseLatency =
-            endTime.getTime() - this.operationStartTime.getTime();
-        }
+    withMetricsDebug(() => {
+      if (!this.firstResponseLatency) {
+        checkState(
+          this.state,
+          [
+            MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_NO_ROWS_YET,
+          ],
+          () => {
+            this.state =
+              MetricsCollectorState.OPERATION_STARTED_ATTEMPT_IN_PROGRESS_SOME_ROWS_RECEIVED;
+            const endTime = new Date();
```

Comment:
same comment about `hrtime.bigint` timestamps

Reply:
This change has been cherry picked.